setup/context/dockerauth.go

```diff
@@ -0,0 +1,116 @@
+package context
+
+// Utilities to load and save docker config files (config.json or dockercfg).
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+const (
+	// CONFIGFILE supplies current docker config file name.
+	CONFIGFILE = "config.json"
+	// OLD_CONFIGFILE supplies old docker config file name.
+	OLD_CONFIGFILE = "dockercfg"
+)
+
+// AuthConfig contains authorization information for connecting to a Registry.
+type AuthConfig struct {
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+	Auth     string `json:"auth,omitempty"`
+
+	// Email is an optional value associated with the username.
+	// This field is deprecated and will be removed in a later
+	// version of docker.
+	Email string `json:"email,omitempty"`
+
+	ServerAddress string `json:"serveraddress,omitempty"`
+
+	// IdentityToken is used to authenticate the user and get
+	// an access token for the registry.
+	IdentityToken string `json:"identitytoken,omitempty"`
+
+	// RegistryToken is a bearer token to be sent to a registry
+	RegistryToken string `json:"registrytoken,omitempty"`
+}
+
+// DockerConfigFile contains `config.json` file info.
+// `dockercfg` files use just the `AuthConfigs` field.
+type DockerConfigFile struct {
+	AuthConfigs map[string]AuthConfig `json:"auths"`
+	HTTPHeaders map[string]string     `json:"HttpHeaders,omitempty"`
+	filename    string                // Note: not serialized - for internal use only
+	Empty       bool                  // Note: not serialized - for internal use only
+}
+
+// LoadDockerConfig loads docker config files of config.json or dockercfg formats.
+func LoadDockerConfig(configDir string) (*DockerConfigFile, error) {
+	if configDir == "" {
+		return nil, fmt.Errorf("No configdir supplied.")
+	}
+
+	configFile := DockerConfigFile{
+		AuthConfigs: make(map[string]AuthConfig),
+		filename:    filepath.Join(configDir, CONFIGFILE),
+	}
+
+	// Try config.json file first
+	if _, err := os.Stat(configFile.filename); err == nil {
```

Comment:
always check `err != nil`, and don't created nested blocks if you can help it. This should be:
```
if _, err := os.Stat(configFile.filename); err != nil {
    return nil, err
}

file, err := os.Open(configFile.filename)
...

```